SepException: distributionWallet is empty

Reference Server is failing with following exception. This issue addresses the bug and adds test coverage to ensure its caught in unit tests.

15:29:19.689 WARN  - o.s.b.w.s.c.AnnotationConfigServletWebServerApplicationContext - Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'uniqueAddressService' defined in URL [jar:file:/app/anchor-platform-runner.jar!/BOOT-INF/lib/anchor-reference-server-0.1.1-plain.jar!/org/stellar/anchor/reference/service/UniqueAddressService.class]: Bean instantiation via constructor failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.stellar.anchor.reference.service.UniqueAddressService]: Constructor threw exception; nested exception is org.stellar.anchor.api.exception.SepException: distributionWallet is empty
15:29:19.689 INFO  - o.s.o.j.LocalContainerEntityManagerFactoryBean - Closing JPA EntityManagerFactory for persistence unit 'default'
15:29:19.888 INFO  - o.s.b.a.l.ConditionEvaluationReportLoggingListener -

Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
15:29:20.090 ERROR - o.s.b.SpringApplication - Application run failed
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'uniqueAddressService' defined in URL [jar:file:/app/anchor-platform-runner.jar!/BOOT-INF/lib/anchor-reference-server-0.1.1-plain.jar!/org/stellar/anchor/reference/service/UniqueAddressService.class]: Bean instantiation via constructor failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.stellar.anchor.reference.service.UniqueAddressService]: Constructor threw exception; nested exception is org.stellar.anchor.api.exception.SepException: distributionWallet is empty
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:315) ~[spring-beans-5.3.15.jar!/:5.3.15]
	at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:296) ~[spring-beans-5.3.15.jar!/:5.3.15]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.autowireConstructor(AbstractAutowireCapableBeanFactory.java:1372) ~[spring-beans-5.3.15.jar!/:5.3.15]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1222) ~[spring-beans-5.3.15.jar!/:5.3.15]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582) ~[spring-beans-5.3.15.jar!/:5.3.15]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542) ~[spring-beans-5.3.15.jar!/:5.3.15]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335) ~[spring-beans-5.3.15.jar!/:5.3.15]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-5.3.15.jar!/:5.3.15]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333) ~[spring-beans-5.3.15.jar!/:5.3.15]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208) ~[spring-beans-5.3.15.jar!/:5.3.15]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:953) ~[spring-beans-5.3.15.jar!/:5.3.15]
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:918) ~[spring-context-5.3.15.jar!/:5.3.15]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:583) ~[spring-context-5.3.15.jar!/:5.3.15]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:145) ~[spring-boot-2.6.3.jar!/:2.6.3]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:732) [spring-boot-2.6.3.jar!/:2.6.3]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:414) [spring-boot-2.6.3.jar!/:2.6.3]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:302) [spring-boot-2.6.3.jar!/:2.6.3]
	at org.stellar.anchor.reference.AnchorReferenceServer.start(AnchorReferenceServer.java:35) [anchor-reference-server-0.1.1-plain.jar!/:?]
	at org.stellar.anchor.platform.ServiceRunner.startAnchorReferenceServer(ServiceRunner.java:64) [classes!/:?]
	at org.stellar.anchor.platform.ServiceRunner.main(ServiceRunner.java:29) [classes!/:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:49) [anchor-platform-runner.jar:?]
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:108) [anchor-platform-runner.jar:?]
	at org.springframework.boot.loader.Launcher.launch(Launcher.java:58) [anchor-platform-runner.jar:?]
	at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:88) [anchor-platform-runner.jar:?]
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.stellar.anchor.reference.service.UniqueAddressService]: Constructor threw exception; nested exception is org.stellar.anchor.api.exception.SepException: distributionWallet is empty
	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:224) ~[spring-beans-5.3.15.jar!/:5.3.15]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:117) ~[spring-beans-5.3.15.jar!/:5.3.15]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:311) ~[spring-beans-5.3.15.jar!/:5.3.15]
	... 27 more
Caused by: org.stellar.anchor.api.exception.SepException: distributionWallet is empty
	at org.stellar.anchor.reference.service.UniqueAddressService.<init>(UniqueAddressService.java:20) ~[anchor-reference-server-0.1.1-plain.jar!/:?]
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[?:?]
	at java.lang.reflect.Constructor.newInstance(Constructor.java:490) ~[?:?]
	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:211) ~[spring-beans-5.3.15.jar!/:5.3.15]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:117) ~[spring-beans-5.3.15.jar!/:5.3.15]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:311) ~[spring-beans-5.3.15.jar!/:5.3.15]
	... 27 more
Reeces-MacBook-Pro~/dev/java-stellar-anchor-sdk %

--- a/anchor-reference-server/src/main/java/org/stellar/anchor/reference/controller/UniqueAddressController.java
+++ b/anchor-reference-server/src/main/java/org/stellar/anchor/reference/controller/UniqueAddressController.java
@@ -1,15 +1,14 @@
 package org.stellar.anchor.reference.controller;
 
-import com.google.gson.Gson;
 import org.springframework.web.bind.annotation.*;
 import org.stellar.anchor.api.callback.GetUniqueAddressResponse;
 import org.stellar.anchor.reference.service.UniqueAddressService;
-import org.stellar.anchor.util.GsonUtils;
+import org.stellar.anchor.util.ConditionalOnPropertyNotEmpty;
 
 @RestController
+@ConditionalOnPropertyNotEmpty("anchor.settings.distributionWallet")
 public class UniqueAddressController {
   final UniqueAddressService uniqueAddressService;
-  static final Gson gson = GsonUtils.builder().create();
 
   public UniqueAddressController(UniqueAddressService uniqueAddressService) {
     this.uniqueAddressService = uniqueAddressService;

--- a/anchor-reference-server/src/main/java/org/stellar/anchor/reference/service/UniqueAddressService.java
+++ b/anchor-reference-server/src/main/java/org/stellar/anchor/reference/service/UniqueAddressService.java
@@ -5,10 +5,12 @@ import org.springframework.stereotype.Service;
 import org.stellar.anchor.api.callback.GetUniqueAddressResponse;
 import org.stellar.anchor.api.exception.SepException;
 import org.stellar.anchor.reference.config.AppSettings;
+import org.stellar.anchor.util.ConditionalOnPropertyNotEmpty;
 import org.stellar.anchor.util.MemoHelper;
 import org.stellar.sdk.KeyPair;
 
 @Service
+@ConditionalOnPropertyNotEmpty("anchor.settings.distributionWallet")
 public class UniqueAddressService {
   String distributionWallet;
   String distributionWalletMemo;

--- a/core/src/main/java/org/stellar/anchor/util/ConditionalOnPropertyNotEmpty.java
+++ b/core/src/main/java/org/stellar/anchor/util/ConditionalOnPropertyNotEmpty.java
@@ -1,0 +1,34 @@
+package org.stellar.anchor.util;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.util.Map;
+import org.jetbrains.annotations.NotNull;
+import org.springframework.context.annotation.Condition;
+import org.springframework.context.annotation.ConditionContext;
+import org.springframework.context.annotation.Conditional;
+import org.springframework.core.type.AnnotatedTypeMetadata;
+
+@Target({ElementType.TYPE, ElementType.METHOD})
+@Retention(RetentionPolicy.RUNTIME)
+@Conditional(ConditionalOnPropertyNotEmpty.OnPropertyNotEmptyCondition.class)
+public @interface ConditionalOnPropertyNotEmpty {
+  String value();
+
+  class OnPropertyNotEmptyCondition implements Condition {
+
+    @Override
+    public boolean matches(@NotNull ConditionContext context, AnnotatedTypeMetadata metadata) {
+      Map<String, Object> attrs =
+          metadata.getAnnotationAttributes(ConditionalOnPropertyNotEmpty.class.getName());
+      if (attrs == null) {
+        return false;
+      }
+      String propertyName = (String) attrs.get("value");
+      String val = context.getEnvironment().getProperty(propertyName);
+      return val != null && !val.trim().isEmpty();
+    }
+  }
+}

